aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -0,0 +1,516 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/core/Tensor.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/cpu/vec/vec.h>
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/native/CPUBlas.h>
+#include <ATen/native/cpu/utils.h>
+#include <ATen/native/transformers/attention.h>
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+#include <c10/util/irange.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/empty.h>
+#endif
+
+#if AT_MKL_ENABLED()
+#include <mkl.h>
+#endif
+
+namespace at::native {
+
+namespace {
+
+using fVec = vec::Vectorized<float>;
+static auto f_vec_size = fVec::size();
+
+inline void _store(
+    float* dst,
+    fVec src,
+    int64_t l=fVec::size()) {
+  src.store(dst, l);
+}
+
+inline void _store(
+    at::BFloat16* dst,
+    fVec src,
+    int64_t l=fVec::size()) {
+  auto res = vec::convert_float_bfloat16(src, src);
+  res.store(dst, l);
+}
+
+inline void _mkl_gemm(
```

Comment:
this shall be gone in future.

cpublas::gemm has some additional checks and normalized the 0 stride (some corner cases when tensor is broadcasted). directly go into mkl would have segfault.